src/main/java/pl/edu/mimuw/BankClient.java

```diff
@@ -1,35 +1,96 @@
 package pl.edu.mimuw;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Random;
 
 public class BankClient {
 
-  // TODO:
-  // - create proper constructor
-  // - implement methods
-  // - add your own methods (1 is enough)
+  private final String firstName;
+  private final String lastName;
+  private final int age;
+  private final String accountNumber;
+  private final List<BankAction> actions;
+
+  public BankClient(String firstName, String lastName, int age) {
+    this.firstName = firstName;
+    this.lastName = lastName;
+    this.age = age;
+
+    var sb = new StringBuilder();
+    var randomGenerator = new Random();
+    for (int i = 0; i < 10; i++) {
+      sb.append(randomGenerator.nextInt(10));
+    }
+    this.accountNumber = sb.toString();
+
+    this.actions = new ArrayList<BankAction>();
+  }
+
+  public BankClient(BankClient client) {
+    this.firstName = client.firstName;
+    this.lastName = client.lastName;
+    this.age = client.age;
+    this.accountNumber = client.accountNumber;
+    this.actions = client.actions;
+  }
 
   public String getName() {
-    throw new IllegalStateException("TODO");
+    return this.firstName + " " + this.lastName;
   }
 
   public String getSurname() {
-    throw new IllegalStateException("TODO");
+    return this.lastName;
   }
 
   public int getAge() {
-    throw new IllegalStateException("TODO");
+    return this.age;
   }
 
   public String getAccountNumber() {
-    throw new IllegalStateException("TODO");
+    return this.accountNumber;
+  }
+
+  public void addAction(BankAction action) {
+    this.actions.add(action);
   }
 
   public List<BankAction> getActionHistory() {
-    throw new IllegalStateException("TODO");
+    final var newActions = new ArrayList<BankAction>();
+
+    for (var action: this.actions) {
+      newActions.add(action.copy());
+    }
+
+    return newActions;
   }
 
-  public void addAction(BankAction action) {
-    throw new IllegalStateException("TODO");
+  public List<BankAction> getActions() {
+    return this.actions;
   }
+
+  protected double sumActions(Class<?> cls) {
```

Comment:
nie robimy takich rzeczy - to świadczy raczej w tym przypadku że źle wyraziliśmy nasz model, jeżeli musimy sprawdzać klasy czy coś takiego. Może wystarczyłoby mieć jednak 2 oddzielne zmienne do loan i deposit, jeżeli chcesz je tak rozdzielać na poziomie intefejsu klienta (jego metod)

Reply:
też by nie zadziałało na dla klas dziedziczących z `DepositBankAction` więc raczej to `instanceOf` w ostateczności, ale też na razie z dala od tego sie trzymamy